docs/proposals/storage_consul_support.md

```diff
@@ -0,0 +1,86 @@
+# Storage Interface Simplification To Support Consul as a KVS
+----------
+
+## Abstract
+-----------
+
+A proposal for a new simplified storage interface with object serialization separated into a higher layer - Raw Storage - in order to reduce the dependencies and requirements for storage backend implementations and to reduce the package dependencies and bootstrap requirements for alternative consumers. The currently existing interface is then implemented by an object serialization layer that wraps around this Raw Storage layer.
+
+## Use Cases
+-----------
+
+1. Allow for alternate KV storage backends to be implemented with less replicated code and complexity.  In our case, Consul support.
+2. Allow consumers like mesos master election, as defined in contrib/mesos/pkg/election, to make use of this Raw Storage to abstract their KV storage backend without adding complexity to bootstrapping or adding excessive package dependencies.
+
+## Motivation
+----------
+
+Many datacenter operators have expressed interest in unifying the KV storage services running on their cluster. (See [#1957](https://github.com/kubernetes/kubernetes/issues/1957)). Usage of EtcD in the code base is not always well-encapsulated, so in order to migrate all EtcD dependencies to an alternate storage solution, some fringe uses of etcd needed to be able to access KV storage without object serialization functionality.
+
+## Raw Storage Implementation
+----------
+
+The RawLayer operates on raw data objects independent of the object serialization layer. These raw objects are little more than a byte array and some consensus flavoring sprinkled on them.
+
+```golang
+type RawObject struct {
+        Data    []byte
+        Version uint64
+        TTL     int64
+}
+```
+
+----------
+Watches and the events they generate also deal exclusively in raw data objects.
+
+```golang
+type RawEvent struct {
+        Type        watch.EventType
+        Current     RawObject
+        Previous    RawObject
+        ErrorStatus interface{}
+}
+type RawWatch interface {
+        Stop()
+        ResultChan() <-chan RawEvent
+}
+```
+----------
+The Raw Storage interface is modeled closely after the existing interface stripped of object serialization requirements and of methods that differ only semantically but not functionally.
+
+```golang
+type RawStorage interface {
+  Create(ctx context.Context, key string, data []byte, raw *RawObject, ttl uint64) error
+
+  Delete(ctx context.Context, key string, raw *RawObject, preconditions RawFilterFunc) error
+
+  Watch(ctx context.Context, key string, resourceVersion string) (RawWatch, error)
```

Comment:
Can you clarify what does resourceVersion mean in this Watch method?

What happens when the resourceVersion is smaller than current resourceVersion in the underlying KV? This would affect how the upper layer should rebuild its in memory state from list + watch.